### ISSUE TITLE
oauth1.signRequest breaks in a certain case

### ISSUE THREAD

**Issue by vivian-duong** [http4s/http4s#3519]
`oauth1.signRequest` does not properly sign a request if the request was created from `Uri.unsafeFromString` with a `?` at the end of the string. I would get a `401 unauthorized` error with the `?`. 

Passing in a uri with a `?` at the end is not super clean because it is not necessary, the `?` just means the start of a query, but as is `Uri.unsafeFromString` didn't give me any problems in the past, it seems like oauth1 should also be able to work with it or users should be encouraged to not include the `?` at the end of a uri passed to `Uri.unsafeFromString`.

I ran the following code and got the following output. Some code was elided.

```
def program(k: String) = BlazeClientBuilder[IO](scala.concurrent.ExecutionContext.global).resource.use {
        client => for {
            uriClean <- IO(org.http4s.Uri.unsafeFromString("https://ads-api.twitter.com/7/stats/accounts/394302/active_entities")
                                    .withQueryParam("entity", "LINE_ITEM")
                                    .withQueryParam("start_time", "2020-03-10")
                                    .withQueryParam("end_time", "2020-05-10"))
            _   <- IO(println(s"\n\n uriClean = $uriClean \n"))
            _   <- IO(println(s"\n\n uriClean.copy(query = org.http4s.Query.empty, fragment = None).renderString = ${uriClean.copy(query = org.http4s.Query.empty, fragment = None).renderString}"))
            reqClean <- IO(Request[IO](method=org.http4s.Method.GET,
                               uri=uriClean))
            _   <- IO(println(s"""\n\n reqClean.uri = ${reqClean.uri} \n"""))
            _   <- IO(println)
            uriDirty   <- IO(org.http4s.Uri.unsafeFromString("https://ads-api.twitter.com/7/stats/accounts/394302/active_entities?")
                                    .withQueryParam("entity", "LINE_ITEM")
                                    .withQueryParam("start_time", "2020-03-10")
                                    .withQueryParam("end_time", "2020-05-10"))
            _       <- IO(println(s"\n\n uriDirty = $uriDirty \n\n"))
            _   <- IO(println(s"\n\n uriDirty.copy(query = org.http4s.Query.empty, fragment = None).renderString = ${uriDirty.copy(query = org.http4s.Query.empty, fragment = None).renderString}"))
            reqDirty <- IO(Request[IO](method=org.http4s.Method.GET,
                               uri=uriDirty))    
            _       <- IO(println(s"\n\n reqDirty.uri = ${reqDirty.uri} \n\n"))
            
            _       <- IO(println(s"getUserParams(reqClean) = ${getUserParams(reqClean)} \n"))
            _       <- IO(println(s"getUserParams(reqDirty) = ${getUserParams(reqDirty)} \n"))
            _       <- IO(println(s"getUserParams(reqClean)==getUserParams(reqDirty) = ${getUserParams(reqClean)==getUserParams(reqDirty)}\n\n"))

        } yield ()
    }
```
```
uriClean = https://ads-api.twitter.com/7/stats/accounts/394302/active_entities?entity=LINE_ITEM&start_time=2020-03-10&end_time=2020-05-10 



 uriClean.copy(query = org.http4s.Query.empty, fragment = None).renderString = https://ads-api.twitter.com/7/stats/accounts/394302/active_entities


 reqClean.uri = https://ads-api.twitter.com/7/stats/accounts/394302/active_entities?entity=LINE_ITEM&start_time=2020-03-10&end_time=2020-05-10 




 uriDirty = https://ads-api.twitter.com/7/stats/accounts/394302/active_entities?&entity=LINE_ITEM&start_time=2020-03-10&end_time=2020-05-10 




 uriDirty.copy(query = org.http4s.Query.empty, fragment = None).renderString = https://ads-api.twitter.com/7/stats/accounts/394302/active_entities


 reqDirty.uri = https://ads-api.twitter.com/7/stats/accounts/394302/active_entities?&entity=LINE_ITEM&start_time=2020-03-10&end_time=2020-05-10 


getUserParams(reqClean) = IO((Request(method=GET, uri=https://ads-api.twitter.com/7/stats/accounts/394302/active_entities?entity=LINE_ITEM&start_time=2020-03-10&end_time=2020-05-10, headers=Headers()),Vector((entity,LINE_ITEM), (start_time,2020-03-10), (end_time,2020-05-10)))) 

getUserParams(reqDirty) = IO((Request(method=GET, uri=https://ads-api.twitter.com/7/stats/accounts/394302/active_entities?&entity=LINE_ITEM&start_time=2020-03-10&end_time=2020-05-10, headers=Headers()),Vector((,), (entity,LINE_ITEM), (start_time,2020-03-10), (end_time,2020-05-10)))) 

getUserParams(reqClean)==getUserParams(reqDirty) = false
```

**Comment by rossabaker** [comment]
I think the essential difference is this (note: if you import `org.http4s.implicits._`, you can use this `uri` syntax instead of the `unsafeFromString`, and it will validate your URIs at compile time):

```scala
scala> uri"https://ads-api.twitter.com/7/stats/accounts/394302/active_entities?".query.toVector
val res20: Vector[(String, Option[String])] = Vector(("",None))

scala> uri"https://ads-api.twitter.com/7/stats/accounts/394302/active_entities".query.toVector
val res21: Vector[(String, Option[String])] = Vector()
```

Maintaining a distinction on the model lets us distinguish the two URIs.  Even though many APIs don't see a difference, some can.

The surprising behavior comes in when we call `withQueryParam`.  We keep that dummy value before adding the new value.  That's how we get the `?&`, which I think is the source of the pain here.

I think we should have represented the query string as an `Option` on the `Uri`, but it's too late to do that for 0.21.  What I think we can do is filter out this dummy empty value when we call withQueryParam, and get a more intuitive behavior.

**Comment by vivian-duong** [comment]
Oh, ok. I didn't know some APIs see a difference between the two URIs. I thought I was just being sloppy with the hanging `?`...?

Thanks for following up.

**Comment by rossabaker** [comment]
Thanks for the report!  I think you can work around it by not dangling the '?', but #3520 will fix even that usage.

**Comment by rossabaker** [comment]
Fixed by #3520, I think.  Please reopen if it still needs attention.
